### ISSUE TITLE
Make tab completion for qvm-, qubes- and other QubesOS-specific commands

### ISSUE THREAD

**Issue by jamke** [QubesOS/qubes-issues#7887]
### The problem you're addressing (if any)
The tab completion for `qvm-`, `qubes-` and other Qubes OS specific commands is not working in Qubes OS.

### The solution you'd like
The proposal is to make it work, providing and completing supported arguments and possible values when possible. 

Examples:

`qvm-ls --run` and `<tab>` => `qvm-ls --running` 
`qvm-ls --t` and `<tab>` => `qvm-ls --tags` 
`qvm-ls --tags` and `<tab><tab>` => shows the list of existing TAGs if possible
`qvm-ls --` and `<tab>` => shows the list of possible arguments the usual bash-way

And support of the list of **names of qubes (VMs)** - would be great! Imagine you type in `qvm-clone `, you press `<tab>` and get not the list of files in the current directory, but the list of VMs suitable for this place:

`qvm-clone ` and `<tab>` => shows the list of existing VMs
`qvm-clone pe` and `<tab>` => `qvm-clone personal`
`qvm-clone sys-` and `<tab><tab>` =>  shows the list of existing VMs that start with `sys-`, because there is more than one.

### Details
* I think it should be a small separate package with a name like `qubes-bash-completion`.
* The list of arguments can be easily taken from docs, `--help` and `man` when available.
* The `bash-completion` package can be used as prototype and and style standard. Here are a lot of examples that cover all cases: https://github.com/scop/bash-completion/tree/master/completions

### Problems
The only downside of such improvement - the completion information should be kept updated with the commands, the same way as documentation and `--help` output now.

### The value to a user, and who that user might be
It will **significantly** improve UX of Qubes OS users who uses terminal at least from time to time (almost everybody, I think).

**Comment by jamke** [comment]
I can try to make it if no one wants.
But it will take a long time and can require my interaction with somebody from the Team.

**Comment by jamke** [comment]
There is already an example of implementation (maybe outdated, that was made 7 years ago by @jgriffiths):
https://github.com/jgriffiths/qubes-completion
His project looks very promising  and interesting.
I am not sure if this particular code should be used in the package, but some ideas for implementation can be definitely taken.

It would be great to include something like that in the dom0 out of the box because otherwise no one will install or even know that it's possible.

**Comment by andrewdavidwong** [comment]
Thank you for the proposal and for volunteering to work on it! Let's see what @marmarek thinks of this idea when he has time to comment.

**Comment by DemiMarie** [comment]
> I can try to make it if no one wants. But it will take a long time and can require my interaction with somebody from the Team.

I would be willing to answer any questions you have.

**Comment by unman** [comment]
I dont think it should take *too* long.

https://github.com/Qubes-Community/Contents/blob/master/code/productivity/qvm-cmds-bash-completion.bash

**Comment by jamke** [comment]
@unman or you meant that you would want to implement completion and it will not take you too long? Maybe I misunderstood.

**Comment by jamke** [comment]
OK, now I'm starting to make an up to date and advanced bash completion for Qubes OS tools as I mentioned above.
I am planning to make it secure and clean enough so it could be  packaged to Qubes OS repos and allow everyone to get more comfort when using Qubes OS out of the box.

> I would be willing to answer any questions you have.

Thank you @DemiMarie, I will contact you (probably by email) when I have questions to ask!

**Comment by jamke** [comment]
I am halfway of making an script for amazing qubes completion.

The completion is already working for **29 Qubes OS tools**, more than a half, and provides completion for everything:
- qube names (VMNAME), taking into account running state and VM type,
- all known arguments, 
- all known flags, 
- options, properties and their possible values,
- prefs,
- features,
- device IDs,
- and more, respecting argument order and -- argument.

A lot of work, including code cleaning, support of volume and pools completions and etc. still has to done.
When all is ready I will provide the script for **testing**.

I also found a bunch of **inconsistencies and minor errors in `man`s of tools and `--help` outputs**.
Do you want me to spend time and make the list of problems?

**The completion also supports different ways of providing options:**
```
qvm-device block attach --option=read-only=yes
qvm-device block attach --option=read-only="yes"
qvm-device block attach --option="read-only=yes"
qvm-device block attach --option="read-only"="yes"

qvm-device block attach --option read-only=yes
qvm-device block attach --option read-only="yes"
qvm-device block attach --option "read-only=yes"
qvm-device block attach --option "read-only"="yes"
```
I checked that all of those ways work in Qubes OS properly and the completion script provides options accordingly. It was hard.


**The list of Qubes OS tools that I decided to make advanced completion for:**
```
qvm-ls
qvm-tags
qvm-start
qvm-shutdown
qvm-kill
qvm-run

qvm-pause
qvm-unpause

qvm-create
qvm-clone
qvm-remove

qvm-device
qvm-block
qvm-usb
qvm-pci

qvm-prefs
qvm-features

qvm-volume

qvm-template
qvm-template-postprocess
qvm-backup
qvm-backup-restore

qvm-check
qvm-firewall
qvm-service
qvm-console-dispvm
qvm-get-image
qvm-get-tinted-image

qvm-pool
qvm-pool-legacy
qvm-sync-appmenus
qvm-appmenus

qvm-copy-to-vm
qvm-move-to-vm
qvm-copy
qvm-move

qvm-start-gui
qvm-start-daemon

qvm-xkill
qvm-sync-clock

qubes-dom0-update
qubes-prefs
qubes-hcl-report
qubesctl
qubesd-query
```

If you think I'm missing something or want to provide any help - let me know.

**Comment by DemiMarie** [comment]
> I am halfway of making a script for amazing qubes completion.
> 
> The completion is already working for **29 Qubes OS tools** (it's more than a half of total amount of selected ones) and provides completion for everything:

Nice!!!

> * qube names (VMNAME), taking into account running state and VM type,
> * all known arguments,
> * all known flags,
> * options, properties and their possible values,
> * prefs,
> * features,
> * device IDs,
> * and more, respecting argument order and -- argument.
> 
> A lot of work, including code cleaning, support of volume and pools completions, still has to be done. When everthing is ready I will provide the script for **testing**.

That would be great!

> I also found a bunch of **inconsistencies and minor errors in `man`s of tools and `--help` outputs**. Do you want me to spend time and make the list of problems?

Yes please!

> **The completion also supports different ways of providing options:**
> 
> ```
> qvm-device block attach --option=read-only=yes
> qvm-device block attach --option=read-only="yes"
> qvm-device block attach --option="read-only=yes"
> qvm-device block attach --option="read-only"="yes"
> 
> qvm-device block attach --option read-only=yes
> qvm-device block attach --option read-only="yes"
> qvm-device block attach --option "read-only=yes"
> qvm-device block attach --option "read-only"="yes"
> ```
> 
> I checked that all of those ways work in Qubes OS properly and the completion script provides options accordingly. It was hard.
> 
> **The list of Qubes OS tools that I decided to make advanced completion for:**
> 
> ```
> qvm-ls
> qvm-tags
> qvm-start
> qvm-shutdown
> qvm-kill
> qvm-run
> 
> qvm-pause
> qvm-unpause
> 
> qvm-create
> qvm-clone
> qvm-remove
> 
> qvm-device
> qvm-block
> qvm-usb
> qvm-pci
> 
> qvm-prefs
> qvm-features
> 
> qvm-volume
> 
> qvm-template
> qvm-template-postprocess
> qvm-backup
> qvm-backup-restore
> 
> qvm-check
> qvm-firewall
> qvm-service
> qvm-console-dispvm
> qvm-get-image
> qvm-get-tinted-image
> 
> qvm-pool
> qvm-pool-legacy
> qvm-sync-appmenus
> qvm-appmenus
> 
> qvm-copy-to-vm
> qvm-move-to-vm
> qvm-copy
> qvm-move
> 
> qvm-start-gui
> qvm-start-daemon
> 
> qvm-xkill
> qvm-sync-clock
> 
> qubes-dom0-update
> qubes-prefs
> qubes-hcl-report
> qubesctl
> qubesd-query
> ```
> 
> If you think I'm missing something or want to provide any help - let me know.

I’d be willing to help review the code.  A few questions:

- What is the latency like?  Slow completion can be worse than no completion, as the shell is unresponsive while the completion code runs.
- For `qubes-dom0-update`, is your code aware of DNF options?

**Comment by jamke** [comment]
> * What is the latency like?  Slow completion can be worse than no completion, as the shell is unresponsive while the completion code runs.

I'm developing it in a qube, not in dom0, so I made simple stubs of qvm-ls, qvm-firewall, qvm-block. And it works instantly, so bash completion script by itself does not slow things down at all. But yes, actual `qvm-*` commands are quite slow (why, btw?), so the latency can be an issue.

> For qubes-dom0-update, is your code aware of DNF options?

Not yet. I'm hoping to somehow pass completion to the `dnf` completion itself (like `sudo` does) as it's quite complicated already. Or I will have to cover some commands myself, like `--enablerepo`.

**Comment by marmarek** [comment]
performance hint: list vms with `qvm-ls --raw-list` or `qvm-ls --raw-data --fields ...` (listing only necessary fields) - it's significantly faster; otherwise it fetches a lot of info for the whole table.

**Comment by jamke** [comment]
> performance hint: list vms with `qvm-ls --raw-list` or `qvm-ls --raw-data --fields ...` (listing only necessary fields) - it's significantly faster; otherwise it fetches a lot of info for the whole table.

@marmarek  yes, thank you, good point. I am already using `--raw-data`, but dropping unused fields (I use the first three columns) is a good idea, especially if it speeds up.

By the way, another possible bug `qvm-ls --help-columns` does not provide `NAME` column, nor is it shown i n`qvm-ls --help`:
```
available columns (see --help-columns):
  CLASS, DISK, FLAGS, GATEWAY, MEMORY, PRIV-CURR, PRIV-MAX, PRIV-USED,
  ROOT-CURR, ROOT-MAX, ROOT-USED, STATE
```

Other issues I see:

1) Not all commands have `--raw` attributes, and parsing output without such flags is not the best idea in the first place. Like output of `ls-pci` is not possible to parse in genera except the first column, because it has a second column of device names that can be anything, including spaces, double spaces and etc.
2) Latency can be a factor. `qvm-ls` runs fast (in --raw-* mode, of course), no problem with that. But running `qvm-volume list` takes 2.7s on my modern desktop PC (whilst `lvs` runs in 0.2s, ten times faster and provides a similar list). Unfortunately, `qvm-volume` has no --raw option, nor I see any way to speed it up as VMNAME is not defined at the point where the completion of volumes is required. That's currently the only place I gave up to complete because of latency is too high for being comfortable.
3) ...

There are many inconsistencies in tools. For example:
- `qvm-ls` supports `--raw-data`, whilst `qvm-firewall` supports `--raw`, some support nothing,
- `qvm-pool` always outputs header line whilst `qvm-device` outputs header depending on output target: terminal and pipe act differently.
- different tools support different sets of default args (--help, --verbose, --quite).

So, I see a room for improvement for all these tools and docs.
I have several proposals how these qvm- and qubes- tools should be addressed to be more consistent and clean. Not sure if the Team would eager to implement it.
Can I modify `man`s myself and provide PRs? Opening issues for each problem or typo is not a good option, probably.

**Comment by DemiMarie** [comment]
> I have several proposals how these qvm- and qubes- tools should be addressed to be more consistent and clean. Not sure if the Team would eager to implement it.

Please make them!

> Can I modify `man`s myself and provide PRs? Opening issues for each problem or typo is not a good option, probably.

Sure!

**Comment by jamke** [comment]
> Please make them!

In @qubes-dev mailing list or on github as a new enhancement issue, where more people sees and discuss it? Or both.

**Comment by andrewdavidwong** [comment]
> On @qubes-dev mailing list or on github as a new enhancement issue, where more people sees and discuss it? Or both.

Sounds like it would be better to start on qubes-devel, then create issue(s) only once concrete, actionable, and well-scoped tasks have been defined.

It's not a matter of where more people see things. The issue tracker and each mailing list has its own rules and purpose. In particular, the issue tracker is not intended to serve as a discussion venue.

**Comment by jamke** [comment]
> * For `qubes-dom0-update`, is your code aware of DNF options?

Update on this - I created a wonderful hack that allows to autocomplete all the `dnf` flags, standalone options and even list packages the same way dnf does.
I reassemble completion command like it's `dnf install ...` and provide completion to `_dnf()`. Similar to what `sudo` does. Actually even better: `dnf install` is replaced with the argument of `--action=` of qubes-dom0-update, so dnf provides flags and other options accordingly. I hope, it will work well.
Plus I avoided the duplication of code from `dnf` completion.

**Comment by jamke** [comment]
> Please make them!

I've posted two proposals in qubes-devel, you probably saw.

P.S. Frankly speaking, I consider qubes-devel to be nearly dead and inactive, so I expect no actual discussions there.
It's understandable as mailing lists from the technical point of view are obsolete, too many disadvantages: no way to make corrections, no proper text formatting, plus no way to login to the web version without google account and etc. And no advantages, as other platforms already do mail notifications with full text and formatting and even allow posting from email if one needs it for some reason (like a habit).
I hope development discussions will someday move to github, gitlab or an least forum, to revive. :)

**Comment by andrewdavidwong** [comment]
@jamke, you are also welcome to post your proposals on the forum, if you like. Unlike qubes-users, qubes-devel is not automatically mirrored on the forum, so you would not be duplicating content that already exists on the forum. (Normally we discourage posting the same thing to multiple mailing lists, but posting to both the qubes-devel and the forum seems fine, as long as the content is appropriate to both venues.)

**Comment by jamke** [comment]
Sorry for delay.
I started using current version in my own `dom0` on daily bases, and hope that when I will be free in a month - I will publish the first version.

**Comment by jamke** [comment]
Still keeping development, currently the project is on stage that I can almost share.

I have a question about `dnf` inside `dom0`. I know that `dom0` is updated via updater qube (e.g. `sys-firewall`), but is it still possible to have bash completion for `dnf`?
Like: 
```
dnf install <TAB><TAB>
```
It does not work inside `dom0` correctly - it provides the list of files in the current directory instead of list of packages (like in any other qubes, even offline ones). How to make it work properly? It would be much more convenient.

**Comment by DemiMarie** [comment]
@jamke I recommend not completing anything for `dnf` in the initial version.

**Comment by jamke** [comment]
> I recommend not completing anything for `dnf` in the initial version.

Myabe I did not get it. I am not talking about `dnf` itself, but `qubes-dom0-update`, of course.
I have implemented a trick with passing completion taskfrom `qubes-dom0-update` to `dnf` . It's like what `sudo` does but more complicated.
And it works like magic, e.g. it provides all flags of `dnf` that can be passed via `qubes-dom0-update` by design.

But when I copy the script from a development qube to `dom0` the completion of packages is not working. I thinks it is because `dnf` has no database in `dom0`. So, I do not know how to fix it, because completion of default `--action=install` would be great.

By the way, completion of packages for commands like: `qubes-dom0-update --action=remove ` does work as expected even in `dom0`, at least this is nice.

**Comment by DemiMarie** [comment]
I suspect this is because the repository used by `qubes-dom0-update has not been generated yet.

**Comment by marmarek** [comment]
Yes, the repository in dom0 has only packages that has been downloaded, there is no full repository metadata. You can probably get it working with `qubes-dom0-update --action=search`, but it would require careful parsing of the (untrusted) output coming from updatevm so IMO it's not worth it.

**Comment by DemiMarie** [comment]
For dom0s that are based on EoL Fedora, what about including the (fixed) Fedora metadata in a package provided by Qubes, and copying the Qubes repository metadata into dom0 (checking the signature in the process)?

**Comment by jamke** [comment]
@DemiMarie  @marmarek I see, that what I thought. Maybe `install` and `search` actions should be indeed addressed some time later. Maybe there will be reasonably secure way to have fixed or even updated metadata of available packages.

Does somebody want to view the code and try the current version, before I clean it a bit more and publish as a beta version on forum or github?

**Comment by jamke** [comment]
> I’d be willing to help review the code.

@DemiMarie, I am ready to share current source code, maybe some recommendations will initiate further changes.
Are you still in and if yes, then how would you prefer to receive it?

I do not want to make it public at current point for a couple of reasons: 1) it may have problems that I am not aware of and people will stuggle, 2) it is not a final version, and if people liked it they may keep this version with possible issues for a while. 3) would like a second opinion before publishing. 
In the future I would love to see it as a part of default installation because it really makes live with Qubes OS much more comfortable, especially for advanced users.

@unman @marmarek maybe you are also interested?

**Comment by unman** [comment]
Of course, interested.
Can you not branch off existing repo, or if it entirely de novo, just
create your own?

**Comment by jamke** [comment]
> Can you not branch off existing repo, or if it entirely de novo, just

It is completely new, not based on any project's source code, only some ideas.

About publishing current version at github - OK, I will work a bit on `README.md` and return shortly with the link.

**Comment by jamke** [comment]
Pushed to `github` and shared the repo with you. I think it should work.
Looking forward for your feedback, guys.

**Comment by jamke** [comment]
Maybe some preliminary feedback is available? 
Should I post it on forum for testing of Qubes OS users (there will be no way back)?

**Comment by jamke** [comment]
Feedback will be appreciated. Even preliminary.

**Comment by jamke** [comment]
Improved completion of `qvm-firewall` and fixed several minor issues. Removing rules is now very pleasant due to table output of `--rule-no=` selector.

**Comment by jamke** [comment]
Preliminary documentation about what the project is able to do is here:
https://forum.qubes-os.org/t/project-qubes-completion-awesome-bash-completion-for-qubes-os-commands/19462

I still would want to make the project public only after preliminary review by somebody from the Team or at least somebody with development skills, but I had no feedback for more than a month, maybe I should publish it as it is.

Nonetheless, anyone who is interested in trying completion and wants to provide feedback - please contact me.

**Comment by andrewdavidwong** [comment]
@jamke: Sorry to hear you haven't received the feedback you wanted. I suppose everyone with the requisite skills has been too busy to volunteer. Since it looks like you're trying to contribute a package, I've updated this issue to conform to our [package contribution guidelines](https://www.qubes-os.org/doc/package-contributions/).

**Comment by jamke** [comment]
@andrewdavidwong I am going to open the repository and hope that it has no major flaws for people who decided to download and use it without reviews.
Preliminary review by somebody with expertise from the Team before making it public is still desired, but it does not look like happening, so be it that way.

**Comment by jamke** [comment]
OK, now anybody can try it!
I made an improved version public, anybody can install it from `Releases` section of github:

https://github.com/jamke/qubes-completion

Currently I would like somebody from the Team make some kind of code review or at least provide some feedback on code quality and information how well it works for you.
Feedback from average users will also be appreciated. So, feel free to try.

**Comment by jamke** [comment]
> I dont believe that you told me what repo was when soliciting review.

Maybe I am not getting you right, but I sent invites on github to the project on May 9. I hoped that github provided information about name of the repo (`qubes-completion`) and the author/owner of repository, so it all will be obvious. Or you mean something else?

**Comment by jamke** [comment]
I changed the first post to reflect that `qubes-completion` [project](https://github.com/jamke/qubes-completion) already exists and ready for use.

**Comment by jamke** [comment]
> Dont have any record of that in email.

I do not know. github shows me that your invite was simply expired after a while. It is strange that you had no email notification for this event. Does not matter now, feel free to try it now.

**Comment by jamke** [comment]
# New release **1.0.2** of `qubes-completion` was published!
Give it a try: https://github.com/jamke/qubes-completion/releases/tag/v1.0.2

### Change log:
* Add support for `qvm-backup` and `qvm-backup-restore`
* Update all basic tests
* Remove third-party code
* Avoid qube completion with `dom0` for some unsuitable cases
* Add submodules of BATS
* Improve README
* Minor typo fixes and improvements

### Installation: 

1. install `bash-completion` package in `dom0`: 
`sudo qubes-dom0-update bash-completion`
2. put bash script-file attached to the release `qubes-completion.sh`  to `/etc/bash_completion.d/` in `dom0`.

More details can be found in the README file.

### Note:

This the release was made with a help of Qubes OS community!

Thanks to all contributors, here is a list of them on github:
https://github.com/jamke/qubes-completion/graphs/contributors

**Comment by jamke** [comment]
I am reminding about this contribution.

It works flawlessly, the Qubes OS experience is really gets much better for advanced user with this clever completion.
And I would really love other people to have it out of the box in the foreseeable future (R4.2 updates, maybe, or at least R4.3).
Can I somehow further help with it?

**Comment by DemiMarie** [comment]
@jamke I think the main problem you are running into is that R4.2 is about to be released, so the developers (especially @marmarek) are busy with the upcoming release.  Once R4.2 is out, the developers will have way more time.

**Comment by DemiMarie** [comment]
@jamke I suggest making a GitHub PR to `qubes-core-admin-client`.  This really should just be part of the default install unless it has significant penalties to performance or attack surface.  It’s a no-brainer from a usability perspective.

**Comment by jamke** [comment]
@DemiMarie well, I use it daily, and it's awesome, but I still don't know who among main devs actually tried my `qubes-completion` project (if any). 
* Have you tried it and use daily it, too?
* Can I consider the project being reviewed and working properly for you, too?
* Any code/security suggestions?

---

> I feel this would be one of the most important usability improvements to qubes that could be made

@ddevz thank you for such kind words.

**Comment by DemiMarie** [comment]
> @DemiMarie well, I use it daily, and it's awesome, but I still don't know who among main devs actually tried my `qubes-completion` project (if any).
> 
> * Have you tried it and use daily it, too?

I have not, because I don’t use contrib packages (except Whonix templates).

> * Can I consider the project being reviewed and working properly for you, too?

If you make a PR I can look at the code.

> * Any code/security suggestions?

See above.

> > I feel this would be one of the most important usability improvements to qubes that could be made
> 
> @ddevz thank you for such kind words.

**Comment by jamke** [comment]
I am here. I will make a PR with this completion feature at the end of this month (currently cannot). 

I have a technical question for discussion: do we want this new feature in the default install as: 
- a single script file in the existing package, or 
- a separate dnf/rpm package like qubes-completion in the qubes os repo?

**Comment by jamke** [comment]
@marmarek , @DemiMarie , @ @andrewdavidwong and other Team members, I would like your opinion on how to add this feature to Qubes OS - as a patch (PR) for existing package or a separate package that user can remove from dom0 it they like? See the last message in this ticket above.

**Comment by jamke** [comment]
> Please see: https://www.qubes-os.org/doc/source-code/#how-to-send-patches

@andrewdavidwong thank you for link, but I do not see how it answers my question.

**Comment by marmarek** [comment]
> as a patch (PR) for existing package or a separate package that user can remove from dom0 it they like? See the last message in this ticket above.

IMO it's more logical to put it together with the tools it's about (so, probably core-admin-client repo). This way, it will be easier to update when relevant tools gain new options.

**Comment by jamke** [comment]
> IMO it's more logical to put it together with the tools it's about (so, probably core-admin-client repo). This way, it will be easier to update when relevant tools gain new options.

OK, patch it is then.

**Comment by jamke** [comment]
Actively working on a project.
Checking, updating and improving all supported commands for R4.2:
https://github.com/jamke/qubes-completion/commits/main/

When it is ready for packaging into the system, I will notify here.

**Comment by jamke** [comment]
# New release **1.1.0-RC1** of `qubes-completion` just published!
Give it a try: https://github.com/jamke/qubes-completion/releases/tag/v1.1.0-rc1

### Change log:

* Full support of R4.2 and all its commands.

* Big code update of `qubes-dom0-update` completion to improve it and make compatible with future dnf completion functions.

* Made the code compatible with the modern `bash-completion` project (up to Fedora 42 and later).

* Added support for new commands of R4.2, like:
	- `qvm-template`, 
	- `qubes-fwupdmgr`, 
	- `qubes-vm-update`, 
	- `qubes-prepare-vm-kernel`, 
	- `qubes-app-menu`,
	- `qubes-update-gui`, 
	- `qubes-policy-lint`,
	- `qubes-policy-editor`.

* A lot of code improvements for existing functions.

### Installation: 

1. install `bash-completion` package in `dom0`: 
`sudo qubes-dom0-update bash-completion`
2. put bash script-file `qubes-completion.sh` that is attached to the release to `/etc/bash_completion.d/` in `dom0`.

More details can be found in the README file.

### Note:

This the release was made with a help of @neoniobium who not only test it, but also provided PRs.

Thanks to all contributors:
https://github.com/jamke/qubes-completion/graphs/contributors

**Comment by jamke** [comment]
OK, I've finished!

The version 1.1.0-RC1 supports ALL commands of R4.2.
It works well for me, and I suggest other people to test it, too.

After 1-2 weeks of test period, if no major issues found, I release v1.1.0. 
After that I will make a GitHub PR to `qubes-core-admin-client` as suggested by @DemiMarie above (https://github.com/QubesOS/qubes-issues/issues/7887#issuecomment-2445074708).

**Comment by ddevz** [comment]
Thanks for doing this!

By the way, did the pull request happen?

**Comment by jamke** [comment]
> Thanks for doing this!

You are welcome, Qubes OS community deserves proper terminal completion.

> By the way, did the pull request happen?

Not yet, I am still waiting for some feedback for this 1.1.0 version, because it has a lot of changes. Still no issues have been reported for it, though.

**Comment by crass** [comment]
> OK, I've finished!

This is a great contribution!

> The version 1.1.0-RC1 supports ALL commands of R4.2. It works well for me, and I suggest other people to test it, too.

I wonder what differences there are with 4.3.

> After 1-2 weeks of test period, if no major issues found, I release v1.1.0. After that I will make a GitHub PR to `qubes-core-admin-client` as suggested by [@DemiMarie](https://github.com/DemiMarie) above ([#7887 (comment)](https://github.com/QubesOS/qubes-issues/issues/7887#issuecomment-2445074708)).

I notice that the completion script is pretty large and that it takes noticeably longer to start a shell with the completions vs without. I wonder if there's anything that can be done about that.

**Comment by jamke** [comment]
> I wonder what differences there are with 4.3.

There are some differences. Due to them I had to postpone PRing upsteam.

> I notice that the completion script is pretty large and that it takes noticeably longer to start a shell with the completions vs without.

I do not notice it in my case, the whole opening of `dom0` xfce terminal takes under a second.
How big is the noticeable delay in your case?

**Comment by crass** [comment]
> > I notice that the completion script is pretty large and that it takes noticeably longer to start a shell with the completions vs without.
> 
> I do not notice it in my case, the whole opening of `dom0` xfce terminal takes under a second. How big is the noticeable delay in your case?

1-3 seconds, whereas without its a fraction of a second.

**Comment by jamke** [comment]
> 1-3 seconds, whereas without its a fraction of a second.

It's not normal. Do you have SSD, right?

I have it under a second for both: with and without completions.

**Comment by jamke** [comment]
@crass  you do not use file `debug_overrides.sh`, do you? Because it turns debug stuff on.

**Comment by neoniobium** [comment]
> > > I notice that the completion script is pretty large and that it takes noticeably longer to start a shell with the completions vs without.
> > 
> > 
> > I do not notice it in my case, the whole opening of `dom0` xfce terminal takes under a second. How big is the noticeable delay in your case?
> 
> 1-3 seconds, whereas without its a fraction of a second.

I have also never experienced this. 

Are you certain that is specific to `qubes-completion` _itself_ or does this also happen if only the normal bash compilation is used? I can notice a small delay (less then 1s) in compression to not using _any_ bash completion.

**Comment by crass** [comment]
> [@crass](https://github.com/crass) you do not use file `debug_overrides.sh`, do you? Because it turns debug stuff on.

Not that I'm aware of.

**Comment by crass** [comment]
> > > > I notice that the completion script is pretty large and that it takes noticeably longer to start a shell with the completions vs without.
> > > 
> > > 
> > > I do not notice it in my case, the whole opening of `dom0` xfce terminal takes under a second. How big is the noticeable delay in your case?
> > 
> > 
> > 1-3 seconds, whereas without its a fraction of a second.
> 
> I have also never experienced this.
> 
> Are you certain that is specific to `qubes-completion` _itself_ or does this also happen if only the normal bash compilation is used? I can notice a small delay (less then 1s) in compression to not using _any_ bash completion.

I'm not sure what "normal bash compilation" entails. The delay is noticed for new shells when the qube completion script is put in `/etc/bash_completion.d` and the delay goes away when that file is removed from there. Nothing else changes.
